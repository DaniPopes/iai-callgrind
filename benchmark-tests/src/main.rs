fn main() {
    println!("{}", fibonacci(30));
    for (key, value) in std::env::vars() {
        println!("{key}={value}");
    }
}

#[inline(never)]
fn fibonacci(n: u64) -> u64 {
    match n {
        0 | 1 => 1,
        n => fibonacci(n - 1) + fibonacci(n - 2),
    }
}
