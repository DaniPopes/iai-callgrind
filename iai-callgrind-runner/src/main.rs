use std::io::Write;

use colored::{control, Colorize};
use env_logger::Env;
use iai_callgrind_runner::{write_all_to_stderr, IaiCallgrindError};
use log::error;
use version_compare::Cmp;

fn main() {
    // Configure the colored crate to respect CARGO_TERM_COLOR
    if let Some(var) = option_env!("CARGO_TERM_COLOR") {
        if var == "never" {
            control::set_override(false);
        } else if var == "always" {
            control::set_override(true);
        }
    }

    // Configure the env_logger crate to respect CARGO_TERM_COLOR
    env_logger::Builder::from_env(
        Env::default()
            .default_filter_or("warn")
            .write_style("CARGO_TERM_COLOR"),
    )
    .format(|buf, record| {
        writeln!(
            buf,
            "{}: {:<5}: {}",
            record
                .module_path()
                .unwrap_or(record.module_path_static().unwrap_or("???")),
            match record.level() {
                log::Level::Error => "Error".red().bold(),
                log::Level::Warn => "Warn".yellow().bold(),
                log::Level::Info => "Info".green().bold(),
                log::Level::Debug => "Debug".blue().bold(),
                log::Level::Trace => "Trace".cyan().bold(),
            },
            record.args()
        )
    })
    .init();

    match iai_callgrind_runner::run() {
        Ok(_) => {}
        Err(error) => {
            match error {
                IaiCallgrindError::VersionMismatch(cmp, runner_version, library_version) => {
                    match cmp {
                        Cmp::Lt => error!(
                            "iai-callgrind-runner ({}) is older than iai-callgrind ({}). Please \
                             update iai-callgrind-runner",
                            runner_version, library_version
                        ),
                        Cmp::Gt => error!(
                            "iai-callgrind-runner ({}) is newer than iai-callgrind ({}). Please \
                             update iai-callgrind",
                            runner_version, library_version
                        ),
                        Cmp::Ne => error!(
                            "No version information found for iai-callgrind but \
                             iai-callgrind-runner ({0}) is >= '0.3.0'. Please update \
                             iai-callgrind to '{0}'",
                            runner_version
                        ),
                        _ => unreachable!(),
                    }
                }
                IaiCallgrindError::LaunchError(exec, error) => {
                    error!("Error executing '{}': {}", exec.display(), error)
                }
                IaiCallgrindError::BenchmarkLaunchError(output) => {
                    error!("Captured stderr:",);
                    write_all_to_stderr(&output.stderr);
                    error!(
                        "Error launching benchmark: Exit code was: {}",
                        output.status.code().unwrap()
                    );
                }
                IaiCallgrindError::Other(message) => {
                    error!("{}", message);
                }
            }
            std::process::exit(1)
        }
    }
}
