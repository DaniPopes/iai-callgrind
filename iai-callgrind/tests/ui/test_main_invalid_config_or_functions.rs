mod test_main_when_no_config_no_function {
    use iai_callgrind::main;
    fn some_func() {}
    main!(functions = );
}

mod test_main_when_wrong_config {
    use iai_callgrind::{main, BinaryBenchmarkConfig};
    fn some_func() {}
    main!(
        config = BinaryBenchmarkConfig::default();
        functions = some_func);
}

mod test_main_when_literal_as_config {
    use iai_callgrind::{main, BinaryBenchmarkConfig};
    fn some_func() {}
    main!(
        config = "my_config";
        functions = some_func);
}

fn main() {}
