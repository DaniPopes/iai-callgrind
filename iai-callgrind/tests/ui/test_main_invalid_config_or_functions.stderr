error: unexpected end of macro invocation
 --> tests/ui/test_main_invalid_config_or_functions.rs:4:22
  |
4 |     main!(functions = );
  |                      ^ missing tokens in macro arguments

warning: unused import: `BinaryBenchmarkConfig`
  --> tests/ui/test_main_invalid_config_or_functions.rs:16:31
   |
16 |     use iai_callgrind::{main, BinaryBenchmarkConfig};
   |                               ^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default
help: consider adding a `#[cfg(test)]` to the containing module
  --> tests/ui/test_main_invalid_config_or_functions.rs:15:1
   |
15 | mod test_main_when_literal_as_config {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `RunnerLibraryBenchmarkConfig: From<BinaryBenchmarkConfig>` is not satisfied
  --> tests/ui/test_main_invalid_config_or_functions.rs:10:5
   |
10 | /     main!(
11 | |         config = BinaryBenchmarkConfig::default();
12 | |         functions = some_func);
   | |______________________________^ the trait `From<BinaryBenchmarkConfig>` is not implemented for `RunnerLibraryBenchmarkConfig`
   |
   = help: the following implementations were found:
             <RunnerLibraryBenchmarkConfig as From<&LibraryBenchmarkConfig>>
             <RunnerLibraryBenchmarkConfig as From<&mut LibraryBenchmarkConfig>>
             <RunnerLibraryBenchmarkConfig as From<LibraryBenchmarkConfig>>
   = note: required because of the requirements on the impl of `Into<RunnerLibraryBenchmarkConfig>` for `BinaryBenchmarkConfig`
   = note: this error originates in the macro `main` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `RunnerLibraryBenchmarkConfig: From<&str>` is not satisfied
  --> tests/ui/test_main_invalid_config_or_functions.rs:18:5
   |
18 | /     main!(
19 | |         config = "my_config";
20 | |         functions = some_func);
   | |______________________________^ the trait `From<&str>` is not implemented for `RunnerLibraryBenchmarkConfig`
   |
   = help: the following implementations were found:
             <RunnerLibraryBenchmarkConfig as From<&LibraryBenchmarkConfig>>
             <RunnerLibraryBenchmarkConfig as From<&mut LibraryBenchmarkConfig>>
             <RunnerLibraryBenchmarkConfig as From<LibraryBenchmarkConfig>>
   = note: required because of the requirements on the impl of `Into<RunnerLibraryBenchmarkConfig>` for `&str`
   = note: this error originates in the macro `main` (in Nightly builds, run with -Z macro-backtrace for more info)
